WiP: Integrate smart contracts and IPFS with frontend

Digital-Mercenaries opened the pull request:
Integrate smart contracts and IPFS with frontend

## Change log (so far) summary

- **`zorp-frontend/package.json`**
  - add IPFS and OpenPGP dependencies
  - add `npm run` script short-cut to turn-off NextJS telemetry
- **`zorp-frontend/next.config.js`**
  - add function to automate copying interface contracts into public assets
- **`zorp-contracts/README.md`**
  - add note for contract consumers
- **`zorp-frontend/src/lib/utils/contractUtils.ts`**
  - add wrappers for initializing JS interfaces from smart-contracts
- **`zorp-frontend/src/lib/utils/ipfsUtils.ts`**
  - add stub wrapper for `verifiedFetch`, expect some churn here
- **`zorp-frontend/src/components/features/UploadIpfs.tsx`**
  - start implementing encrypted upload to IPFS functionality!
  - `6617163` has more detailed check-list of things that need sorted

If you've got pointers on the NextJS/ReactJS stuff, specifically on how shared state/context should be handled, then I'd much appreciate it!...  There's a lot of doc-churn in this space, and I certainly don't trust the LLMs with any data safety stuff ;-)

Digital-Mercenaries committed:
Add NPM script to opt-out of NextJS telemetry...

> Why is not default behavior?

Digital-Mercenaries committed:
Ignore artifacts built by `zorp-contracts`...

These _should_ be built via;

```bash
cd zorp-contracts
forge build
```

...  And then copied by some Next JS or CI/CD thang to a path served to
the public, eg. `zorp-frontend/public/assets/...`

Digital-Mercenaries committed:
Sketch-out some NextJS integration customization

> Yes those that cry for DRY are likely to be annoyed by this (-;

Digital-Mercenaries committed:
Note for third-parties seeking interface contracts

Digital-Mercenaries committed:
Add client-side dependencies

- `@helia/verified-fetch` _should_ provide IPFS fetch
- `openpgp` _should_ provide GPG/PGP encryption

Digital-Mercenaries committed:
Add wrappers to fetch and parse Zorp ABIs

> Note; error messages probably should be adjusted to be more
> informative, and type-hints probably would benefit from more hinting.

Digital-Mercenaries committed:
Add stub wrapper for IPFS fetching

Digital-Mercenaries committed:
Add OpenPGP wrapper for encrypting data

> Note; UX/UI may wanna make it really obvious that data is encrypted
> client-side, I.E. before anything hits a network or chain!

Digital-Mercenaries committed:
:warning: Add partial IPFS upload implementation

**Thangs required to complete**

- [ ] GPG/PGP key from `ZorpStudy.encryptionKey()` IPFS CID
- [ ] GPG upload from client stored in session/context
- [ ] IPFS pinning API integration
- [ ] IPFS upload confirmation and CID retrieval or computing
- [ ] `getZorpStudy` from `src/lib/utils/contractUtils.ts` shared state
- [ ] Blockchain submission via `IZorpStudy.submitData()`

wkyleg commented on zorp-frontend/src/lib/utils/contractUtils.ts:
```diff
@@ -0,0 +1,74 @@
+	name: string;
+	address: `0x${string}`;
+	client: Client /* | KeyedClient */;
+}) {
```
This is main thing we will want to change. Better is to pull in the contracts with WAGMI, that way the types will be wrapped with ABI Type too. Essentially, WAGMI wraps ABI type to extract valid type signatures from the contract. You do need to same the contract ABI in a TS file however.

https://wagmi.sh/react/guides/write-to-contract

Digital-Mercenaries replied:
:facepalm: hah, I totally spaced we can import `.json` files now!

I'll remove the overly complex code and use more React-ish stuff in the next set of changes :-)

Digital-Mercenaries committed:
Add alias for contract ABIs

Digital-Mercenaries committed:
:warning: Partially implement expected actions...

High-level outline of what _should_ happen:

0. Get data from `<input>` file upload
1. Get encryption public key from client
2. Get encryption public key from `ZorpStudy.encryptionKey()`
3. Encrypt file data with OpenPGP JS
4. Upload encrypted file to IPFS
5. Write IPFS pointer to blockchain via `ZorpStudy.submitData(ipfs_cid)`

Where friction is felt most:

- React wrappers `useReadContract` and `useWriteContract` from `wagmi`
   > ... smells of ye-olde XMLHttpRequest call-back hell
- IPFS upload/pinning, solving will require either;
  - backend/lambda/clown-runner
  - or study creators to host IPFS pinning/auth infra
  - or a reasonably mature DeFi to IPFS solution
  - or integration with Filecoin, which would add to client requirements

Digital-Mercenaries committed:
Make peace with React... _mostly_...

:warning: the `throw new Error('TODO: Upload to IPFS')` on line `154` is
intentional to prevent blockchain interactions until a solution for
pinning is sorted.

:crossed_fingers: Otherwise things _should_ just work, if that is
relevant documentation is correct and correctly understood ;-)

Digital-Mercenaries committed:
Add tips/tricks for local testing

> Note; will need an IPFS test runner to join this party for full
> end-to-end testing, and it wouldn't hurt to have a `scripts/`
> directory at the repo root for semi-automating things.

wkyleg reviewed (approved):
Looks good! We will need to add in a few loading states with the contracts but this is great progress

Digital-Mercenaries commented:
Super!...  Yeah I figure the UI/UX should be sorted after we've got the execution pipeline, and states, fully figured out.
